### ISSUE TITLE
Espresso alternative for UIAutomator locator

### ISSUE THREAD

**Issue by ganjarpanji** [appium/appium-espresso-driver#499]
Currently im using so many elements with uiautomator locator :

> "new UiSelector().resourceId(\"com.myapp.android.staging:id/info_top_text_view\").text(\"today\")")

wondering if espresso has native locator similar to this instead of using Xpath?. 
And i tried search the documentation of the espresso-driver locator but not no luck, only found the datamatcher.

Thanks

**Comment by KazuCocoa** [comment]
The UiSelector is not Espresso's feature. So espresso driver also does not support it.
Espresso has `viewtag` and `datamatcher` so far.

**Comment by TarCV** [comment]
@ganjarpanji view matchers (based on Hamcrest) are used to locate views in Espresso tests, see https://developer.android.com/training/testing/espresso/basics#finding-view
For example, your locator would be written like `onView(allOf(withId(R.id.info_top_text_view), withText("today")))` in an Espresso test.
But AFAIK view matchers are not supported by Appium Espresso driver.

**Comment by ganjarpanji** [comment]
really looking forward this to be implemented :)

> onView(allOf(withId(R.id.info_top_text_view), withText("today"))) in an Espresso test.

**Comment by ganjarpanji** [comment]
hi @KazuCocoa its been quite progress for appium-espresso. confirming regarding this issue seems now it's supported by the driver?

**Comment by KazuCocoa** [comment]
viewmatcher and datamatcher are supported now. They work like https://github.com/appium/appium-espresso-driver/issues/499#issuecomment-538408266
e.g. https://github.com/appium/appium-espresso-driver/blob/f7a0495facf2d41812e26cfd3cd36dc40a56a310/test/functional/commands/find-e2e-specs.js
(Please find proper Espresso methods in their documentation when you implement `viewmatcher` strategy)

`resourceId` works as `id` lookup strategy, but not all UIAutomator methods work on espresso driver since the backend is Espresso. You should replace each method to proper Espresso methods via `viewmatcher` etc or `id` etc lookup strategies.

They should be enough as espresso driver, I assume.

----

Basically, we do not have any plans to add calling UIAutomator methods via this espresso driver to keep Espresso benefits.
(I forgot where, but I also saw an issue report that UIAutomator causes some issues when combining it with Espresso. So such combination could lead to another potential issue)

**Comment by ganjarpanji** [comment]
thank you so much, this would be enough. I'm planning to migrate my whole tests script to espresso driver soon. closing issue
